active/0024-opentelemetry-trace-integration.md

```diff
@@ -0,0 +1,214 @@
+
+> Messaging systems themselves may trace messages as the messages travels from producers to consumers. Such tracing would cover the transport layer but would not help in correlating producers with consumers. To be able to directly correlate producers with consumers, another context that is propagated with the message is required.
+>
+> A message creation context allows correlating producers with consumers of a message and model the dependencies between them, regardless of the underlying messaging transport mechanism and its instrumentation.
+>
+> The message creation context is created by the producer and should be propagated to the consumer(s).
+>
+> A producer SHOULD attach a message creation context to each message. If possible, the message creation context SHOULD be attached in such a way that it cannot be changed by intermediaries.
+
+In fact, EMQX is capable of participating in distributed trace out of the box (without OpenTelemetry instrumentation), simply because it implements the above MQTT specification requirement and propagates User-Properties from a publisher to a subscriber.
+
+However, if no trace context received but a message still should be traced, one of the following options should be chosen:
+
+- create internal trace context and trace only internal EMQX events and do not propagate the context to receivers and/or external data systems (if bridges are set up)
+- create internal trace context and propagate it to receivers and/or external data systems.
+
+The option can be made configurable and should probably default to not propagating internally created trace context (**TBD**). 
+
+### Attributes (TBD)
+
+OpenTelemetry defines [some conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/messaging-spans.md#messaging-attributes) (status: Experimental at the time of writing this document).
+
+The attributes are grouped under several name-spaces:
+
+- `messaging.*`
+- `network.*`
+- `server.*`
+
+It should be  decided whether to follow these conventions strictly or define our own (e.g. mqtt.topic, mqtt.clientid etc).
+Another open question is whether to always include all the attributes or make them configurable, so that a user may select appropriate ones from the pre-defined list.
+
+### Sampling
+
+OpenTelemetry sampling is described in great depth in the [official documentation](https://opentelemetry.io/docs/concepts/sampling).
+
+Erlang opentelemetry lib implements only head sampling. Head sampling implies that a sampling decision is made as early as possible, e.g., by following a configured percentage of traces to sample (100% by default). A decision to sample or drop a span or trace is not made by inspecting the trace as a whole.
+
+Sampling rate option should be added to EMQX configuration.
+
+Tail sampling that makes a sampling decision after all the spans are done would need to be implemented by extending opentelemetry lib (**TBD**).
```

Comment:
The examples of tail sampling below sound more like they are a form of filtering rather than sampling.  Perhaps those are good candidates for filter criteria, since they are not quite random?

Reply:
Head/tail sampling is OpenTelemetry terminology.
Tail sampling is essentially stateful: it depends on the given trace events, conditions occurred some time after the trace started. 
Filtering is EMQX specific, inspired by EMQX log tracing, it must stateless and quite simple (if need it at all). 
I agree that both concepts are very similar, especially when sampling means non-probability based approaches.
Probably, they should be somehow unified but we also need to define which ones need to be implemented (at least in the first iteration).